# 45장. 프로미스

## 프로미스가 있기 전에

### 콜백 헬

비동기 함수는 비동기 처리 결과를 외부에 반환할 수 없고, 상위 스코프의 변수에 할당할 수도 없다. 따라서 비동기 함수의 처리결과에 대한 후속 처리는 콜백 함수를 전달하는 것이 일반적이다.  
그러나 콜백 함수를 통해 비동기 처리 결과에 대한 후속 처리를 수행하는 비동기 함수가 비동기 처리 결과를 가지고 또다시 비동기 함수를 호출해야 한다면 콜백 함수 호출이 중첩되어 복잡도가 높아지는 현상이 발생하는데, 이를 **콜백 헬**이라 한다. 콜백 헬은 가독성을 나쁘게 하며 실수를 유발하는 원인이 된다.

```js
get("/step1", (a) => {
  get(`/step2/${a}`, (b) => {
    get(`/step3/${b}`, (c) => {
      get(`/step4/${c}`, (d) => {
        console.log(d);
      });
    });
  });
});
```

<br>

### 에러처리의 한계

```js
try {
  setTimeout(() => {
    throw new Error("Error!");
  }, 1000);
} catch (e) {
  // 에러를 캐치하지 못한다
  console.error("캐치한 에러", e);
}
```

비동기 함수인 `setTimeout`이 호출되면 `setTimeout` 함수의 실행 컨텍스트가 생성되어 콜 스택에 푸시되어 실행된다. `setTimeout`은 비동기 함수이므로 콜백 함수가 호출되는 것을 기다리지 않고 즉시 종료되어 콜 스택에서 제거된다.  
`setTimeout` 함수의 콜백 함수가 실행될 때 `setTimeout` 함수는 이미 콜 스택에서 제거된 상태다. 따라서 에러 `setTimeout` 함수의 콜백 함수가 발생시킨 에러는 `catch` 블록에서 캐치되지 않는다.

<br>

## 프로미스

`Promise` 생성자 함수를 `new` 연산자와 함께 호출하면 프로미스를 생성한다. ES6에서 도입된 `Promise`는 호스트 객체가 아닌 ECMAScript 사양에 정의된 표준 빌트인 객체다.  
`Promise` 생성자 함수는 비동기 처리를 수행할 콜백 함수를 인수로 전달받는데 이 콜백 함수는 `resolve`와 `reject` 함수를 인수로 전달받는다.

```js
// 프로미스 생성
const Promise = new Promise((resolve, reject) => {
   // Promise 함수의 콜백 함수 내부에서 비동기 처리를 수행한다.
   if(/* 비동기 처리 성공 */) {
      resolve('result');
   } else { /* 비동기 처리 실패 */
      reject('failure reason');
   }
});
```

<br>

프로미스는 다음과 같이 현재 비동기 처리가 어떻게 진행되고 있는지를 나타내는 상태 정보를 갖는다.

| 프로미스의 상태 정보 | 의미                                  | 상태 변경 조건                   |
| -------------------- | ------------------------------------- | -------------------------------- |
| `pending`            | 비동기 처리가 아직 수행되지 않은 상태 | 프로미스가 생성된 직후 기본 상태 |
| `fulfilled`          | 비동기 처리가 수행된 상태 (성공)      | `resolve` 함수 호출              |
| `rejected`           | 비동기 처리가 수행된 상태 (실패)      | `reject` 함수 호출               |

생성된 직후의 프로미스는 기본적으로 `pending` 상태다 이후 비동기 처리가 수행되면 비동기 처리 결과에 따라 `fulfilled` 또는 `rejected` 상태로 변경되고 이와 상관 없이 `pending`이 아닌 상태로 비동기 처리가 수행된 상태를 `settled` 상태라고 한다.
<br>

## 프로미스의 후속 처리 메서드

### `Promise.prototype.then`

`then` 메서드는 두개의 콜백 함수를 인수로 전달받는다. 첫 번째 콜백 함수는 비동기 처리가 성공했을 때 호출되는 성공 처리 콜백 함수이며, 두 번째 콜백 함수는 비동기 처리가 실패했을 때 호출되는 실패 처리 콜백 함수다.

```js
// fulfilled
new Promise((resolve) => resolve("fulfilled")).then(
  (v) => console.log(v),
  (e) => console.error(e)
); // fulfilled

// rejected
new Promise((_, reject) => reject(new Error("rejected"))).then(
  (v) => console.log(v),
  (e) => console.error(e)
); // Error: rejected
```

`then` 메서드는 언제나 프로미스를 반환한다. 만약 `then` 메서드의 콜백 함수가 프로미스를 반환한다면 그 프로미스를 그대로 반환하고, 프로미스가 아닌 값을 반환하면 그 값을 암묵적으로 `resolve` 또는 `reject`하여 프로마스를 생성해 반환한다.

### `Promise.prototype.catch`

`catch` 메서드는 한개의 콜백 함수를 인수로 전달받는다. `catch` 메서드의 콜백 함수는 프로미스가 `rejected`상태인 경우만 호출된다.

```js
// rejected
new Promise((_, reject) => reject(new Error("rejected"))).catch((e) => console.log(e)); // Error: rejected
```

`catch` 메서드는 `then`과 동일하게 동작한다. 따라서 `then` 메서드와 마찬가지로 언제나 프로미스를 반환한다.

```js
// rejected
new Promise((_, reject) => reject(new Error("rejected"))).then(undefined, (e) => console.log(e)); // Error: rejected
```

### `Promise.prototype.finally`

`finally` 메서드는 한 개의 콜백 함수를 인수로 전달받는다. `finally` 메서드의 콜백 함수는 프로미스의 성공 또는 실패와 상관없이 무조건 한 번 호출된다. `finally` 메서드도 `then`/`catch` 메서드와 마찬가지로 언제나 프로미스를 반환한다.

```js
new Promise(() => {}).finally(() => console.log("finally")); // finally
```

<br>

## 프로미스의 에러 처리

프로미스의 비동기 처리 결과에 대한 후속 처리는 프로미스가 제공하는 후속 처리 메서드 `then`, `catch`, `finally`를 사용하여 수행한다.

```js
const wrongUrl = "https://jsonplaceholder.typicode.com/XXX/1";

promiseGet(wrongUrl).then(
  (res) => console.log(res),
  (err) => console.error(err)
);
```

단, `then` 메서드의 두번째 콜백 함수는 첫번째 콜백 함수에서 발생한 에러를 캐치하지 못하고 가독성이 떨어져 좋지 않기 때문에 에러처리는 `catch` 메서드에서 하는 것을 권장한다.

```js
const wrongUrl = "https://jsonplaceholder.typicode.com/XXX/1";

promiseGet(wrongUrl)
  .then((res) => console.log(res))
  .catch((err) => console.error(err));
```

<br>

## 프로미스의 정적 메서드

### `Promise.resolve` / `Promise.reject`

인수로 전달 받은 값을 resolve/reject 하는 프로미스를 생성한다.

### `Promise.all`

여러개의 비동기 처리를 모두 병렬 처리할 때 사용한다.  
프로미스를 요소로 갖는 배열 등의 이터러블을 인수로 전달 받고, 전달받은 모든 프로미스가 모두 `fulfilled` 상태가 되면 모든 처리 결과를 배열에 저장해 새로운 프로미스를 반환한다.  
이때 `fulfilled` 상태가 되는 순서와 상관 없이 처리 순서가 보장 된다.

```js
const requestData1 = () => new Promise((resolve) => setTimeout(() => resolve(1), 3000));
const requestData2 = () => new Promise((resolve) => setTimeout(() => resolve(2), 2000));
const requestData3 = () => new Promise((resolve) => setTimeout(() => resolve(3), 1000));

Promise.all([requestData1(), requestData2(), requestData3()])
  .then(console.log) // [1, 2, 3] => 약 3초 소요
  .catch(console.error);
```

만약 전달받은 배열의 프로미스가 하나라도 `rejected` 상태가 되면 나머지 프로미스가 `fulfilled` 상태가 되는 것을 기다리지 않고 즉시 종료한다.

```js
Promise.all([
  new Promise((_, reject) => setTimeout(() => reject(new Error("Error 1")), 3000)),
  new Promise((_, reject) => setTimeout(() => reject(new Error("Error 2")), 2000)),
  new Promise((_, reject) => setTimeout(() => reject(new Error("Error 3")), 1000)),
])
  .then(console.log)
  .then(console.error); // Error: Error 3
```

### `Promise.race`

`Promise.race` 메서드는 `Promise.all` 메서드와 동일하게 프로미스를 요소로 갖는 배열 등의 이터러블을 인수로 전달받는다.  
`Promise.race` 메서드는 `Promise.all` 메서드처럼 모든 프로미스가 `fulfilled` 상태가 되는 것을 기다리는 것이 아니라 가장 먼저 `fulfilled` 상태가 된 프로미스의 처리 결과를 `resolve`하는 새로운 프로미스를 반환한다.

```js
Promise.race([
  new Promise((resolve) => setTimeout(() => resolve(1), 3000)),
  new Promise((resolve) => setTimeout(() => resolve(2), 2000)),
  new Promise((resolve) => setTimeout(() => resolve(3), 1000)),
])
  .then(console.log) // 3
  .catch(console.error);
```

프로미스가 `rejected` 상태가 되면 `Promise.all` 메서드와 동일하게 처리된다.

### `Promise.allSettled`

전달받은 프로미스가 모두 `settled` 상태가 되면 처리 결과를 배열로 반환한다. ES11에 도입된 `Promise.allSettled` 메서드는 IE를 제외한 대부분의 모던 브라우저에서 지원한다.

```js
Promise.allSettled([
  new Promise((resolve) => setTimeout(() => resolve(1), 2000)),
  new Promise((_, reject) => setTimeout(() => reject(new Error("Error!")), 1000)),
]).then(console.log);

/*
[
  {status: "fulfilled", value: 1},
  {status: "rejected", reason: Error: Error! at <anonymous>:3:54},
]
*/
```

<br>

## 마이크로태스크 큐

```js
setTimeout(() => console.log(1), 0);

Promise.resolve()
  .then(() => console.log(2))
  .then(() => console.log(3));
```

위의 예시는 1 -> 2 -> 3 순으로 출력될 것 처럼 보이지만 2 -> 3 -> 1 순으로 출력된다. 그 이유는 프로미스의 후속 처리 메서드의 콜백 함수는 태스크 큐가 아니라 마이크로태스크 큐에 저장되기 때문이다. 마이크로태스크 큐는 태스크 큐와는 별도의 큐이며, 마이크로태스크 큐가 태스크큐보다 우선순위가 높다. 즉, 이벤트 루프는 콜스택이 비면 먼저 마이크로태스크 큐에서 대기하고 있는 함수를 가져와 실행하고, 이후 마이크로태스크 큐가 비면 태스크 큐에서 대기하고 있는 함수를 가져와 실행한다.
